### PR TITLE
New GPU binaries with cuQuantum support

### PR DESCRIPTION
**Pull request by doichanj**
### Summary
Change supported CUDA version from 10 to 11.2 or higher and support cuQuantum SDK.

### Details and comments
Now there are separate binaries for CUDA 11 and CUDA 12:

`qiskit-aer-gpu` for CUDA 12 and
`qiskit-aer-gpu-cu11` for CUDA 11 for backward compatibility

required libraries are automatically installed via pip for each CUDA version.
Thanks to the demo https://github.com/NVIDIA/cuQuantum/tree/main/extra/demo_build_with_wheels

### CODE REVIEW

**Comment by hhorii** [comment]
> Now there are separate binaries for CUDA 11 and CUDA 12:
> 
> qiskit-aer-gpu-cu11 for CUDA 11, and
> qiskit-aer-gpu-cu12 for CUDA 12
> 
> required libraries are automatically installed via pip for each CUDA version.

Will we deprecate qiskit-aer-gpu?

File: README.md [hunk 1]
```diff
@@ -29,6 +29,11 @@ If you want to install our GPU supported simulators, you have to install this ot
 pip install qiskit-aer-gpu
 ```
 
+The package avobe is for CUDA&reg 12, so if your system has CUDA&reg; 11 installed, install separate package:
```
Comment (mtreinish) [inline README.md:32]:
```suggestion
The package above is for CUDA&reg 12, so if your system has CUDA&reg; 11 installed, install separate package:
```

File: releasenotes/notes/renew_gpu_binaries-2cf3eba0853b8407.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+---
+upgrade:
+  - |
+    Changed supported CUDA version for GPU simulator to 11.2 or higher.
+    Now GPU binaries are including cuQuantum support. There are separate
+    binaries for CUDA 11 and 12. qiskit-aer-gpu-cu11 for CUDA 11 and
+    qiskit-aer-gpu-cu12 for CUDA 12.
```
Comment (mtreinish) [inline releasenotes/notes/renew_gpu_binaries-2cf3eba0853b8407.yaml:7]: I think it's critical we document the upgrade path for existing users of `qiskit-aer-gpu`  here. Users already have enough difficulty navigating the gpu support in aer for whatever reason, so I think we need to explicitly document in this release note how you go from `qiskit-aer-gpu` to `qiskit-aer-gpu-cu11`  or `qiskit-aer-gpu-cu12` and also put this in the README too.

File: .github/workflows/deploy.yml [hunk 3]
```diff
@@ -119,28 +119,35 @@ jobs:
...
           TWINE_USERNAME: qiskit
         run: twine upload dist/qiskit*
-  gpu-build:
-    name: Build qiskit-aer-gpu wheels
+  gpu-build-cuda11:
+    name: Build qiskit-aer-gpu-cu11 wheels
     runs-on: ubuntu-latest
     steps:
+      - name: Maximize build space
+        uses: easimon/maximize-build-space@master
```
Comment (mtreinish) [inline .github/workflows/deploy.yml:127]: Are we running out of space with the newer cuda builds?

File: .github/workflows/deploy.yml [hunk 4]
```diff
@@ -119,28 +119,35 @@ jobs:
...
       - name: Build wheels
         env:
-          CIBW_BEFORE_ALL: "yum install -y yum-utils wget && wget -q https://developer.download.nvidia.com/compute/cuda/10.1/Prod/local_installers/cuda-repo-rhel6-10-1-local-10.1.243-418.87.00-1.0-1.x86_64.rpm && rpm -i cuda-repo-rhel6-10-1-local-10.1.243-418.87.00-1.0-1.x86_64.rpm && yum clean all && yum -y install cuda-10-1 openblas-devel"
-          CIBW_MANYLINUX_X86_64_IMAGE: quay.io/pypa/manylinux2010_x86_64:latest
-          CIBW_SKIP: "*-manylinux_i686 cp310* pp* cp36* *musllinux*"
-          CIBW_TEST_SKIP: "cp*"
-          CIBW_ENVIRONMENT: QISKIT_AER_PACKAGE_NAME=qiskit-aer-gpu AER_THRUST_BACKEND=CUDA CUDACXX=/usr/local/cuda/bin/nvcc
+          CIBW_BEFORE_ALL: "yum install -y yum-utils wget && wget -q https://developer.download.nvidia.com/compute/cuda/11.8.0/local_installers/cuda-repo-rhel7-11-8-local-11.8.0_520.61.05-1.x86_64.rpm && rpm -i cuda-repo-rhel7-11-8-local-11.8.0_520.61.05-1.x86_64.rpm && yum clean all && yum -y install cuda && yum -y install openblas-devel && yum-config-manager --add-repo https://developer.download.nvidia.com/compute/cuda/repos/rhel7/x86_64/cuda-rhel7.repo && yum clean all"
```
Comment (mtreinish) [inline .github/workflows/deploy.yml:146]: Have you done a test to validate this works in the buildwheel env? We've had a lot of issues in the past with the nvidia packages not actually setting up the compilers correctly.

File: setup.py [hunk 5]
```diff
@@ -20,6 +20,50 @@
...
+]
+
+if "gpu" in PACKAGE_NAME:
+    if "11" in CUDA_MAJOR:
+        requirements_cuda = [
+            "nvidia-cuda-runtime-cu11>=11.8.89",
+            "nvidia-cublas-cu11>=11.11.3.6",
+            "nvidia-cusolver-cu11>=11.4.1.48",
+            "nvidia-cusparse-cu11>=11.7.5.86",
+            "cuquantum-cu11>=23.3.0",
```
Comment (mtreinish) [inline setup.py:48]: So these nvidia packages include the dynamic library files for all the cuda symbols we need which is why you exclude them from the repair wheel command?

File: releasenotes/notes/renew_gpu_binaries-2cf3eba0853b8407.yaml [hunk 6]
```diff
@@ -0,0 +1,7 @@
+---
+upgrade:
+  - |
+    Changed supported CUDA version for GPU simulator to 11.2 or higher.
```
Comment (mtreinish) [inline releasenotes/notes/renew_gpu_binaries-2cf3eba0853b8407.yaml:4]:
```suggestion
    Changed supported CUDA version for GPU simulator to 11.2 or higher. Previously qiskit-aer supported running
    with CUDA>=10.1, but now the minimum requirement is 11.2. This was required to leverage newer features in
    CUDA along with cuQuantum support. It also reflects changes made in Python packaging around the use of
    CUDA.
```

**Review by mtreinish — COMMENTED** [review]
I'm fine with that as a general approach, but I think the most important thing here is what the upgrade path is for existing users of qiskit-aer-gpu. I think we need a release note and more documentation on how an existing qiskit-aer-gpu user will upgrade after this change.  I think the answer is as easy as `pip uninstall qiskit-aer-gpu && pip install -U qiskit-aer-gpu-cu11`, but we need to test this and make it very clear what manual steps are needed.

File: .github/workflows/deploy.yml (hunk 3 above)
Comment (doichanj) [reply .github/workflows/deploy.yml:127]: yes I observed out of space error without this setting. I tested with some parameters here and I set the safety reserved space

File: .github/workflows/deploy.yml (hunk 4 above)
Comment (doichanj) [reply .github/workflows/deploy.yml:146]: Here is repository I used for tests. https://github.com/doichanj/qiskit-aer/tree/lite_GPU_package
And I tested installing and running qiskit-aer-gpu from testpypi uploaded by this one

File: setup.py (hunk 5 above)
Comment (doichanj) [reply setup.py:48]: Yes this is the trick to decrease the binary size

File: releasenotes/notes/renew_gpu_binaries-2cf3eba0853b8407.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (doichanj) [reply releasenotes/notes/renew_gpu_binaries-2cf3eba0853b8407.yaml]: I updated release note

**Review by hhorii — APPROVED** [review]

**Pull request merged**